### PR TITLE
Http signature dereferencing

### PR DESCRIPTION
**Pull request by lamara**

### CODE REVIEW

File: lib/HttpSignatureStrategy.js [hunk 1]
```diff
@@ -78,12 +81,14 @@ Strategy.prototype.authenticate = function(req) {
             requiredHeaders: REQUIRED_HEADERS
           }));
       }
-      callback();
+      callback(null);
```
Comment (dlongley) [inline lib/HttpSignatureStrategy.js]: Why did you add `null` here? We usually do the simplest thing, which is to call `callback();` with no params, which will set `err` to undefined. There have been some places in various libs that people have done `callback(null);` but I think those are now proliferating when they should be replaced with `callback();`. If there's not a good reason to pass the extra parameter, don't.

File: tests/10-http-signatures.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+var helpers = require('./helpers');
+var brIdentity = require('bedrock-identity');
+var database = require('bedrock-mongodb');
+var mockData = require('./mock.data');
+var store = require('bedrock-credentials-mongodb').provider;
+var uuid = require('node-uuid').v4;
+var url = require('url');
+var util = bedrock.util;
+var request = require('request');
+request = request.defaults({json: true, strictSSL: false});
```
Comment (mattcollier) [inline tests/10-http-signatures.js:16]: Remove unused vars.

File: lib/HttpSignatureStrategy.js [hunk 3]
```diff
@@ -211,13 +216,46 @@ function getDidPublicKey(publicKey, callback) {
...
+  async.auto({
+    dereferencePublicKey: function(callback) {
+      request.get(publicKey, function(err, res, body) {
+        callback(err, res.body);
+      });
+    },
+    dereferenceOwner: ['dereferencePublicKey', function(callback, results) {
+      var dereferencedKey = results.dereferencePublicKey;
+      var owner = dereferencedKey.owner;
+      request.get(owner, function(err, res, body) {
```
Comment (dlongley) [inline lib/HttpSignatureStrategy.js]: Options for `request` should include accept header with: `application/ld+json; application/json`. Otherwise endpoint might return HTML, by default, etc.

File: lib/HttpSignatureStrategy.js (hunk 1 above)
Comment (lamara) [inline lib/HttpSignatureStrategy.js]: will remove!

File: lib/HttpSignatureStrategy.js [hunk 4]
```diff
@@ -11,7 +11,10 @@ var passport = require('passport');
 var util = require('util');
 var BedrockError = bedrock.util.BedrockError;
 var didio = require('did-io');
+var jsonld = require('jsonld');
 var URL = require('url');
+var request = require('request');
+request = request.defaults({json: true, strictSSL: false});
```
Comment (mattcollier) [inline lib/HttpSignatureStrategy.js]: strictSSL should be an option or parameterized somehow.  Here's how we did it in bedrock-messages-client: https://github.com/digitalbazaar/bedrock-messages-client/blob/master/lib/client.js#L42

File: lib/HttpSignatureStrategy.js [hunk 5]
```diff
@@ -211,13 +216,46 @@ function getDidPublicKey(publicKey, callback) {
...
+    dereferencePublicKey: function(callback) {
+      request.get(publicKey, function(err, res, body) {
+        callback(err, res.body);
+      });
+    },
+    dereferenceOwner: ['dereferencePublicKey', function(callback, results) {
+      var dereferencedKey = results.dereferencePublicKey;
+      var owner = dereferencedKey.owner;
+      request.get(owner, function(err, res, body) {
+        callback(err, res.body);
```
Comment (mattcollier) [inline lib/HttpSignatureStrategy.js:260]: Should this just be 'body'?

File: lib/HttpSignatureStrategy.js [hunk 6]
```diff
@@ -211,13 +216,46 @@ function getDidPublicKey(publicKey, callback) {
...
-    'URL scheme "https" is not supported.',
-    'HttpSignature.UnsupportedUrlScheme', {
-      'public': true,
-      httpStatusCode: 400,
-      scheme: 'https'
-    }));
+  async.auto({
+    dereferencePublicKey: function(callback) {
+      request.get(publicKey, function(err, res, body) {
+        callback(err, res.body);
```
Comment (mattcollier) [inline lib/HttpSignatureStrategy.js:232]: Should this just be 'body'?

File: lib/HttpSignatureStrategy.js [hunk 7]
```diff
@@ -211,13 +216,46 @@ function getDidPublicKey(publicKey, callback) {
...
+    }],
+    verify: ['dereferenceOwner', function(callback, results) {
+      var dereferencedOwner = results.dereferenceOwner;
+      var dereferencedKey = results.dereferencePublicKey;
+      var ownerKeys = jsonld.getValues(dereferencedOwner, 'publicKey');
+      var found = false;
+      ownerKeys.forEach(function(key) {
+        if(key.id === dereferencedKey.id) {
+          found = true;
+          return false;
```
Comment (dlongley) [inline lib/HttpSignatureStrategy.js]: Just a note that `return false;` doesn't actually break the `forEach` loop or anything. You'll need to use (misuse?) `.some` to break a loop like this -- or just use the traditional `for(var i = 0; i < ownerKeys.length; ++i) { ... break`

File: lib/HttpSignatureStrategy.js [hunk 8]
```diff
@@ -211,13 +216,46 @@ function getDidPublicKey(publicKey, callback) {
...
-    }));
+  async.auto({
+    dereferencePublicKey: function(callback) {
+      request.get(publicKey, function(err, res, body) {
+        callback(err, res.body);
+      });
+    },
+    dereferenceOwner: ['dereferencePublicKey', function(callback, results) {
+      var dereferencedKey = results.dereferencePublicKey;
+      var owner = dereferencedKey.owner;
```
Comment (mattcollier) [inline lib/HttpSignatureStrategy.js:237]: Seems like there should be checking here to make sure 'owner' exists and maybe that it's a URL before sending it into request.

File: tests/10-http-signatures.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+      async.auto({
+        before: function(callback) {
+          // Remove keys from database to force a dereference on the key url
+          helpers.removeCollection('publicKey', callback);
+        },
+        authenticate: ['before', function(callback) {
+          var clonedUrlObj = util.clone(urlObj);
+          request.get(
+              helpers.createHttpSignatureRequest(
+                  url.format(clonedUrlObj), user),
```
Comment (mattcollier) [inline tests/10-http-signatures.js]: Fix indent in this area.

File: tests/helpers.js [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    httpSignature: {
+      key: identity.keys.privateKey.privateKeyPem,
+      keyId: identity.keys.publicKey.id,
+      headers: ['date', 'host', 'request-line']
+    }
+  };
+  return newRequest;
+};
+
+// FIXME: Don't use truecred as host.
```
Comment (mattcollier) [inline tests/helpers.js]: Don't make refrences to tc in bedrock modules.

File: tests/10-http-signatures.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+var urlObj = {
+  protocol: 'https',
+  host: config.server.host,
+  pathname: '/tests/bedrock-passport/http-signature-test'
+};
+
+describe('bedrock-passport', function() {
+  before('Prepare the database', function(done) {
+    helpers.prepareDatabase(mockData, done);
```
Comment (mattcollier) [inline tests/10-http-signatures.js]: Do you want/need identities inserted into the database?

File: tests/10-http-signatures.js [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+  });
+  after('Remove test data', function(done) {
+    helpers.removeCollections(done);
+  });
+  describe('authenticated requests using http-signature', function() {
+    var user = mockData.identities.mock;
+    it('dereference lookup completes successfully', function(done) {
+      async.auto({
+        before: function(callback) {
```
Comment (mattcollier) [inline tests/10-http-signatures.js]: This is already being handle in 'before', maybe you need a 'beforeEach' as well.

File: tests/helpers.js [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+
+ 'use strict';
+
+var _ = require('lodash');
+var async = require('async');
+var bedrock = require('bedrock');
+var brKey = require('bedrock-key');
+var brIdentity = require('bedrock-identity');
+var config = require('bedrock').config;
+var uuid = require('node-uuid').v4;
```
Comment (mattcollier) [inline tests/helpers.js]: remove unused vars.

File: tests/10-http-signatures.js (hunk 11 above)
Comment (lamara) [inline tests/10-http-signatures.js]: For this specific test -- no, I don't think it's needed, but any other tests are going to want to be able to work with identities that are in the DB.

File: lib/HttpSignatureStrategy.js (hunk 3 above)
Comment (mattcollier) [inline lib/HttpSignatureStrategy.js]: This is being handle by setting json:true on L17.

File: lib/HttpSignatureStrategy.js (hunk 6 above)
Comment (lamara) [inline lib/HttpSignatureStrategy.js:232]: Yes!

File: tests/10-http-signatures.js (hunk 12 above)
Comment (lamara) [inline tests/10-http-signatures.js]: This specific test requires that we remove the database data, but any subsequent test might want it in, so just added it to this test.

File: tests/10-http-signatures.js (hunk 11 above)
Comment (mattcollier) [inline tests/10-http-signatures.js]: Then those tests should probably be in a different describe block or file.

File: tests/10-http-signatures.js (hunk 12 above)
Comment (mattcollier) [inline tests/10-http-signatures.js]: Again, tests that need that arrangement should have their own describe block.

File: lib/HttpSignatureStrategy.js (hunk 5 above)
Comment (dlongley) [inline lib/HttpSignatureStrategy.js:260]: Or maybe not bother with third param `body`. One or the other, I'd imagine.

File: lib/HttpSignatureStrategy.js (hunk 3 above)
Comment (dlongley) [inline lib/HttpSignatureStrategy.js]: I didn't even see that line up there, I'll leave comments on it. We don't want to be using library-wide defaults in a system module like this.

File: lib/HttpSignatureStrategy.js (hunk 4 above)
Comment (dlongley) [inline lib/HttpSignatureStrategy.js]: Note that "json: true" isn't quite what we want here. We want to actually request `application/ld+json; application/json`. That means we have a preference for JSON-LD first, but we'll take JSON if that's not available.

File: lib/HttpSignatureStrategy.js (hunk 3 above)
Comment (dlongley) [inline lib/HttpSignatureStrategy.js]: Nevermind, `request.defaults` gives you a new instance of request, so that's ok. We just need to be more specific about the MIME type we want --- it isn't just JSON.

File: lib/HttpSignatureStrategy.js (hunk 4 above)
Comment (lamara) [inline lib/HttpSignatureStrategy.js]: should strictSSL just be a config option here, then? It seems like it would either always be on or off, because any arbitrary person can have us initiate a request to their publicKey and owner documents. @mattcollier

File: lib/HttpSignatureStrategy.js (hunk 8 above)
Comment (lamara) [inline lib/HttpSignatureStrategy.js:237]: Yes -- need to check that 'owner' is actually an https url!

**Comment by lamara** [comment]
Addressed most of the comments -- will probably want to add more tests though, especially non-happy path tests.
